compiler/controllers.py

```diff
@@ -0,0 +1,281 @@
+"""Request controllers."""
+
+from typing import Tuple, Optional, Callable, Any
+from http import HTTPStatus as status
+
+from werkzeug import MultiDict
+from werkzeug.exceptions import BadRequest, NotFound, InternalServerError, \
+    Forbidden
+
+from flask import url_for
+
+from arxiv.users.domain import Session
+from arxiv.base import logging
+
+from .services import Store, filemanager
+from .services.store import DoesNotExist
+from . import compiler
+from .domain import Task, Product, Status, Format
+
+logger = logging.getLogger(__name__)
+
+Response = Tuple[dict, int, dict]
+
+
+def _status_from_store(source_id: str, checksum: str,
+                       output_format: Format) -> Optional[Task]:
+    """Get a :class:`.Task` from storage."""
+    store = Store.current_session()
+    try:
+        stat = store.get_status(source_id, checksum, output_format)
+        logger.debug('Got status from store: %s', stat)
+        return stat
+    except DoesNotExist as e:
+        logger.debug('No such compilation: %s', e)
+    # except Exception as e:
+    #     logger.debug('No such compilation: %s', e)
+    return None
+
+
+def _redirect_to_status(source_id: str, checksum: str, output_format: Format,
+                        code: int = status.SEE_OTHER) -> Response:
+    """Redirect to the status endpoint."""
+    location = url_for('api.get_status', source_id=source_id,
+                       checksum=checksum, output_format=output_format.value)
+    return {}, code, {'Location': location}
+
+
+def service_status(*args: Any, **kwargs: Any) -> Response:
+    """Exercise dependencies and verify operational status."""
+    fm = filemanager.FileManager.current_session()
+    store = Store.current_session()
+    response_data = {}
+    response_data['store'] = store.is_available()
+    response_data['compiler'] = compiler.is_available()
+    response_data['filemanager'] = fm.is_available()
+    if not all(response_data.values()):
+        return response_data, status.SERVICE_UNAVAILABLE, {}
+    return response_data, status.OK, {}
+
+
+def compile(request_data: MultiDict, token: str, session: Session,
+            is_authorized: Callable = lambda task: True) -> Response:
+    """
+    Start compilation of an upload workspace.
+
+    Parameters
+    ----------
+    request_data : :class:`.MultiDict`
+        Data payload from the request.
+    token : str
+        Auth token to be used for subrequests (e.g. to file management
+        service).
+
+    Returns
+    -------
+    dict
+        Response data.
+    int
+        HTTP status code.
+    dict
+        Headers to add to response.
+
+    """
+    if 'output_format' in request_data:
+        output_format = Format(request_data['output_format'])
```

Comment:
Consider handling the exception in the case where an invalid `output_format` is provided, and raising `BadRequest`.

Reply:
Fixed in ef39993; also updated the other cases where this is handled in ``controllers.py`` to use the same error message.

```python
        try:
            output_format = Format(request_data['output_format'])
        except ValueError as e:
            raise BadRequest(f'Unsupported format: {output_format}') from e
```